subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/WatchingVirtualFileSystem.java

```diff
@@ -175,28 +179,35 @@ public void beforeBuildFinished(boolean watchingEnabled) {
      */
     private synchronized void startWatching() {
         if (watchRegistry != null) {
+            if (!eventConsumerThread.isAlive()) {
+                throw new RuntimeException("The thread consuming the file events stopped for an unknown reason");
+            }
             return;
         }
         try {
             long startTime = System.currentTimeMillis();
             watchRegistry = watcherRegistryFactory.startWatcher(new FileWatcherRegistry.ChangeHandler() {
                 @Override
                 public void handleChange(FileWatcherRegistry.Type type, Path path) {
-                    try {
-                        fileEvents.put(new FileEvent(path, type));
-                    } catch (InterruptedException e) {
-                        Thread.currentThread().interrupt();
-                        throw new RuntimeException(e);
-                    }
+                    addToQueueOrReportOverflow(FileEvent.changed(path, type));
                 }
 
                 @Override
                 public void handleLostState() {
-                    try {
-                        fileEvents.put(new FileEvent(true));
-                    } catch (InterruptedException e) {
-                        Thread.currentThread().interrupt();
-                        throw new RuntimeException(e);
+                    addToQueueOrReportOverflow(FileEvent.lostState());
+                }
+
+                private void addToQueueOrReportOverflow(FileEvent fileEvent) {
+                    boolean addedToQueue = fileEvents.offer(fileEvent);
+                    if (!addedToQueue) {
+                        LOGGER.warn("Couldn't handle file events fast enough, dropping state");
+                        fileEvents.clear();
+                        try {
+                            fileEvents.put(FileEvent.lostState());
+                        } catch (InterruptedException e) {
+                            Thread.currentThread().interrupt();
```

Comment:
It is unclear which thread is interrupted here. Is it the native thread? If so, I don't think we should interrupt that one (or if it has any effect).

Reply:
This does not interrupt the thread, it clears the interrupted flag.

Reply:
I'm confused. Isn't that done via `interrupted()`?

Reply:
You are right. I am now using the right call.